File: python-src/feature_engineering.py
"""
Feature engineering — transform raw records into ML-ready features.

This module builds:
- Course-level features: statistics per (course_name × major)
- Student-level features: per-student context (IPK, IPS, trend)

Status: IMPLEMENTED in TypeScript (src/lib/gpa/feature-engineering.ts)
        and Python (mini-services/ml-service/src/train_models.py)
        This is a stub for the original pipeline structure.
"""

import pandas as pd
import numpy as np


def build_course_features(df: pd.DataFrame) -> pd.DataFrame:
    """
    Aggregate per-course statistics for Model 1 (K-Means clustering).

    Features: mean_gp, std_gp, pct_a_or_above, pct_below_b, pct_fail,
              median_gp, iqr_gp, mean_ips_takers, sample_count
    """
    grouped = df.groupby(["course_name", "major"])

    features = grouped["grade_point"].agg(["mean", "std", "median"]).reset_index()
    features.columns = ["course_name", "major", "mean_gp", "std_gp", "median_gp"]

    features["pct_a_or_above"] = grouped.apply(
        lambda g: (g["grade_point"] >= 3.7).mean()
    ).reset_index(drop=True)

    features["pct_below_b"] = grouped.apply(
        lambda g: (g["grade_point"] < 3.0).mean()
    ).reset_index(drop=True)

    features["pct_fail"] = grouped.apply(
        lambda g: (g["grade_point"] < 1.0).mean()
    ).reset_index(drop=True)

    features["iqr_gp"] = grouped["grade_point"].apply(
        lambda x: x.quantile(0.75) - x.quantile(0.25)
    ).reset_index(drop=True)

    features["mean_ips_takers"] = grouped["grade_point"].mean().reset_index(drop=True)
    features["sample_count"] = grouped.size().reset_index(drop=True)
    features = features.fillna(0)

    return features


def build_student_features(df: pd.DataFrame) -> pd.DataFrame:
    """
    Build per-student×course feature rows for Model 2 (grade prediction).

    Features: difficulty_score, ipk_before, ips_last_semester, ips_trend,
              total_sks_completed, course_sks, semester_number, etc.
    """
    rows = []

    # Build course features for difficulty lookup
    course_feats = build_course_features(df)
    course_lookup = {}
    for _, row in course_feats.iterrows():
        key = (row["course_name"], row["major"])
        course_lookup[key] = row

    # Group by student
    for student_id, student_df in df.groupby("student_id"):
        student_df = student_df.sort_values("semester_number")
        semesters = sorted(student_df["semester_number"].unique())

        ips_by_sem = {}
        cum_sks = 0
        cum_weighted = 0
        ipk_before_by_sem = {}

        for sem in semesters:
            sem_df = student_df[student_df["semester_number"] == sem]
            sks = sem_df["sks"].sum()
            weighted = (sem_df["sks"] * sem_df["grade_point"]).sum()
            ips = weighted / sks if sks > 0 else 0
            ips_by_sem[sem] = ips
            ipk_before_by_sem[sem] = cum_weighted / cum_sks if cum_sks > 0 else ips
            cum_sks += sks
            cum_weighted += weighted

        ips_list = list(ips_by_sem.values())

        for _, r in student_df.iterrows():
            key = (r["course_name"], r["major"])
            cf = course_lookup.get(key)
            if cf is None:
                continue

            sem = r["semester_number"]
            ips_last = ips_by_sem.get(sem, ips_list[-1] if ips_list else 3.3)
            ips_prev = ips_list[ips_list.index(ips_last) - 1] if len(ips_list) > 1 and sem in ips_list else ips_last
            ips_trend = ips_last - ips_prev
            ipk_before = ipk_before_by_sem.get(sem, ips_last)
            total_sks_before = sum(
                student_df[student_df["semester_number"] < sem]["sks"].sum()
                for _ in [0]
            )

            difficulty_score = max(0, min(1,
                (1 - cf["mean_gp"] / 4) * 0.55 + cf["pct_fail"] * 0.25 + cf["pct_below_b"] * 0.2
            ))

            rows.append({
                "difficulty_score": difficulty_score,
                "ipk_before": round(ipk_before, 3),
                "ips_last_semester": round(ips_last, 3),
                "ips_trend": round(ips_trend, 3),
                "total_sks_completed": total_sks_before,
                "course_sks": r["sks"],
                "semester_number": sem,
                "mean_ips_takers": cf["mean_ips_takers"],
                "pct_a_or_above": cf["pct_a_or_above"],
                "pct_below_b": cf["pct_below_b"],
                "grade_class": r["grade"],
            })

    return pd.DataFrame(rows)

File: src/lib/gpa/feature-engineering.ts
/**
 * Feature engineering — transforms raw student/course records into the feature
 * matrices consumed by Model 1 and Model 2.
 * Mirrors Python `src/features/feature_engineering.py`.
 */
import { GRADE_ORDER, MAJORS } from "./config";
import type { CourseFeatures, StudentCourseFeatures } from "./types";
import type { SyntheticRecord } from "./training-data";
import { computeStats } from "./grade-utils";

export interface RawRecord {
  studentId?: string;
  major?: string;
  semesterNumber: number;
  courseName: string;
  sks: number;
  grade: string;
  gradePoint: number;
}

/** Build per-course features grouped by (courseName × major). */
export function buildCourseFeatures(records: RawRecord[]): CourseFeatures[] {
  const groups = new Map<string, RawRecord[]>();
  for (const r of records) {
    const major = (r.major || "Umum").trim();
    const key = `${r.courseName}|||${major}`;
    if (!groups.has(key)) groups.set(key, []);
    groups.get(key)!.push(r);
  }
  const out: CourseFeatures[] = [];
  for (const [key, recs] of groups) {
    const [courseName, major] = key.split("|||");
    const gps = recs.map((r) => r.gradePoint);
    const ips = recs.map((r) => r.gradePoint); // proxy: per-student IPS = gradePoint for this course
    const stats = computeStats(gps);
    const meanIps = computeStats(ips).mean;
    out.push({
      courseName,
      major,
      meanGradePoint: Math.round(stats.mean * 100) / 100,
      stdGradePoint: Math.round(stats.std * 100) / 100,
      pctAOrAbove: Math.round(stats.pctAOrAbove * 100) / 100,
      pctBelowB: Math.round(stats.pctBelowB * 100) / 100,
      pctFail: Math.round(stats.pctFail * 100) / 100,
      medianGradePoint: Math.round(stats.median * 100) / 100,
      iqrGradePoint: Math.round(stats.iqr * 100) / 100,
      meanIpsTakers: Math.round(meanIps * 100) / 100,
      sampleCount: recs.length,
    });
  }
  return out;
}

/**
 * Build student×course feature rows for Model 2 training.
 * Each record is expanded with rolling IPS/IPK context prior to that semester.
 */
export function buildStudentFeatures(records: RawRecord[]): StudentCourseFeatures[] {
  // Group by student
  const byStudent = new Map<string, RawRecord[]>();
  for (const r of records) {
    const sid = r.studentId ?? "anon";
    if (!byStudent.has(sid)) byStudent.set(sid, []);
    byStudent.get(sid)!.push(r);
  }

  const majorEncoding = encodeMajors(MAJORS as readonly string[]);
  const rows: StudentCourseFeatures[] = [];

  for (const [sid, recs] of byStudent) {
    // Sort by semester
    recs.sort((a, b) => a.semesterNumber - b.semesterNumber);
    // Per-semester aggregates
    const semesterAgg = new Map<number, { sks: number; weightedGp: number }>();
    for (const r of recs) {
      if (!semesterAgg.has(r.semesterNumber)) {
        semesterAgg.set(r.semesterNumber, { sks: 0, weightedGp: 0 });
      }
      const agg = semesterAgg.get(r.semesterNumber)!;
      agg.sks += r.sks;
      agg.weightedGp += r.sks * r.gradePoint;
    }
    const semesters = [...semesterAgg.keys()].sort((a, b) => a - b);
    const ipsBySem = new Map<number, number>();
    let cumSks = 0;
    let cumWeighted = 0;
    const ipkBeforeSem = new Map<number, number>();
    for (const s of semesters) {
      const agg = semesterAgg.get(s)!;
      const ips = agg.sks > 0 ? agg.weightedGp / agg.sks : 0;
      ipsBySem.set(s, ips);
      ipkBeforeSem.set(s, cumSks > 0 ? cumWeighted / cumSks : ips);
      cumSks += agg.sks;
      cumWeighted += agg.weightedGp;
    }

    // Course features lookup (per-course × major stats) — recompute on the fly
    const courseFeats = buildCourseFeatures(recs);
    const cfLookup = new Map<string, CourseFeatures>();
    for (const cf of courseFeats) cfLookup.set(`${cf.courseName}|||${cf.major}`, cf);

    for (const r of recs) {
      const cf = cfLookup.get(`${r.courseName}|||${r.major ?? "Umum"}`);
      if (!cf) continue;
      const ipsList = semesters.map((s) => ipsBySem.get(s)!);
      const ipsLast = ipsBySem.get(r.semesterNumber) ?? ipsList[ipsList.length - 1] ?? 3.3;
      const ipsPrev = ipsList[ipsList.indexOf(ipsLast) - 1] ?? ipsLast;
      const ipsTrend = Math.round((ipsLast - ipsPrev) * 100) / 100;
      const ipkBefore = ipkBeforeSem.get(r.semesterNumber) ?? ipsLast;
      const totalSksBefore = semesters
        .filter((s) => s < r.semesterNumber)
        .reduce((sum, s) => sum + (semesterAgg.get(s)?.sks ?? 0), 0);
      // Difficulty features: derive a simple difficulty score from course stats
      const difficultyScore = deriveDifficultyScore(cf);
      const difficultyCluster = difficultyClusterOf(difficultyScore);
      const gradeClass = GRADE_ORDER.indexOf(r.grade);

      // Each scenario adds one training row with a different target distribution
      // (shifted). To keep training simple, we just replicate with scenario=serius.
      rows.push({
        difficultyScore,
        difficultyCluster,
        ipkBefore: Math.round(ipkBefore * 100) / 100,
        ipsLastSemester: Math.round(ipsLast * 100) / 100,
        ipsTrend,
        totalSksCompleted: totalSksBefore,
        courseSks: r.sks,
        semesterNumber: r.semesterNumber,
        scenarioEncoded: 1, // serius — neutral scenario for training
        majorEncoded: majorEncoding[r.major ?? "Umum"] ?? 0,
        meanIpsTakers: cf.meanIpsTakers,
        pctAOrAbove: cf.pctAOrAbove,
        pctBelowB: cf.pctBelowB,
        gradeClass,
      });
    }
    void sid;
  }

  return rows;
}

/** Encode majors to integers. */
export function encodeMajors(majors: readonly string[]): Record<string, number> {
  const map: Record<string, number> = {};
  majors.forEach((m, i) => (map[m] = i));
  map["Umum"] = majors.length;
  return map;
}

/** Derive a 0..1 difficulty score from course features (heuristic, used as bootstrap). */
export function deriveDifficultyScore(cf: CourseFeatures): number {
  // Higher mean → easier → lower difficulty.
  const meanFactor = 1 - cf.meanGradePoint / 4; // 0..1
  const failFactor = cf.pctFail; // 0..1
  const belowB = cf.pctBelowB; // 0..1
  const score = meanFactor * 0.55 + failFactor * 0.25 + belowB * 0.2;
  return Math.max(0, Math.min(1, Math.round(score * 100) / 100));
}

export function difficultyClusterOf(score: number): number {
  if (score < 0.2) return 0;
  if (score < 0.4) return 1;
  if (score < 0.6) return 2;
  if (score < 0.8) return 3;
  return 4;
}
